libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -958,65 +958,77 @@ void CompilerUtils::convertType(
 			// Copy the array to a free position in memory, unless it is already in memory.
 			if (typeOnStack.location() != DataLocation::Memory)
 			{
-				// stack: <source ref> (variably sized)
-				unsigned stackSize = typeOnStack.sizeOnStack();
-				ArrayUtils(m_context).retrieveLength(typeOnStack);
-
-				// allocate memory
-				// stack: <source ref> (variably sized) <length>
-				m_context << Instruction::DUP1;
-				ArrayUtils(m_context).convertLengthToSize(targetType, true);
-				// stack: <source ref> (variably sized) <length> <size>
-				if (targetType.isDynamicallySized())
-					m_context << u256(0x20) << Instruction::ADD;
-				allocateMemory();
-				// stack: <source ref> (variably sized) <length> <mem start>
-				m_context << Instruction::DUP1;
-				moveIntoStack(2 + stackSize);
-				if (targetType.isDynamicallySized())
+				if (
+					typeOnStack.dataStoredIn(DataLocation::CallData) &&
+					typeOnStack.isDynamicallyEncoded() &&
+					typeOnStack.baseType()->isDynamicallyEncoded()
+				)
 				{
-					m_context << Instruction::DUP2;
-					storeInMemoryDynamic(*TypeProvider::uint256());
-				}
-				// stack: <mem start> <source ref> (variably sized) <length> <mem data pos>
-				if (targetType.baseType()->isValueType())
-				{
-					copyToStackTop(2 + stackSize, stackSize);
-					ArrayUtils(m_context).copyArrayToMemory(typeOnStack);
+					solAssert(m_context.useABICoderV2(), "");
+					// stack: <source ref> (variably sized) <length>
```

Comment:
Is the comment here correct? If yes, why can you use a fixed `swap1` below?

Reply:
For dynamically sized I need to swap `length` and `source ref`. Comment is copied from this functions, few lines before

Reply:
But source ref might be a multi-slot variable.

Reply:
Can you add some assertions about the stack sizes?

Reply:
I will add assertions. 
But, shouldn't multi-slot variable be when array is dynamically sized in calldata? Or there is some other case as well?